x/staking/keeper/msg_server.go

```diff
@@ -605,6 +606,77 @@ func (k msgServer) UpdateParams(ctx context.Context, msg *types.MsgUpdateParams)
 	return &types.MsgUpdateParamsResponse{}, nil
 }
 
-func (k msgServer) RotateConsPubKey(_ context.Context, _ *types.MsgRotateConsPubKey) (*types.MsgRotateConsPubKeyResponse, error) {
-	return nil, nil
+func (k msgServer) RotateConsPubKey(goCtx context.Context, msg *types.MsgRotateConsPubKey) (res *types.MsgRotateConsPubKeyResponse, err error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	pk, ok := msg.NewPubkey.GetCachedValue().(cryptotypes.PubKey)
+	if !ok {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidType, "expecting cryptotypes.PubKey, got %T", pk)
```

Comment:
Trying to infer the type from the failed assertion won't give you the right result per https://go.dev/play/p/LcHy1v4rfUb. The correct way to do it is

```go
cv := msg.NewPubKey.GetCachedValue()
pk, ok := cv.(cryptotypes.)
if !ok {
   ..., cv)
}
```